src/aor.cpp

```diff
@@ -0,0 +1,749 @@
+       ++s)
+  {
+    if (s->second->_expires < _next_expires)
+    {
+      _next_expires = s->second->_expires;
+    }
+  }
+
+  // If nothing has altered the _next_expires, the AoR is empty and invalid.
+  // Return 0 to indicate there is nothing to expire.
+  if (_next_expires == INT_MAX)
+  {
+    return 0;
+  }
+  // Otherwise we return the value found.
+  return _next_expires;
+}
+
+void AoR::copy_aor(const AoR& source_aor)
+{
+  for (Bindings::const_iterator i = source_aor.bindings().begin();
+       i != source_aor.bindings().end();
+       ++i)
+  {
+    Binding* src = i->second;
+    Binding* dst = get_binding(i->first);
+    *dst = *src;
+  }
+
+  for (Subscriptions::const_iterator i = source_aor.subscriptions().begin();
+       i != source_aor.subscriptions().end();
+       ++i)
+  {
+    Subscription* src = i->second;
+    Subscription* dst = get_subscription(i->first);
+    *dst = *src;
+  }
+
+  _associated_uris = AssociatedURIs(source_aor._associated_uris);
+  _notify_cseq = source_aor._notify_cseq;
+  _timer_id = source_aor._timer_id;
+  _uri = source_aor._uri;
+  _scscf_uri = source_aor._scscf_uri;
+}
+
+// Patch the AoR. This does the following:
+//
+// TODO
+void AoR::patch_aor(const PatchObject& po)
+{
+  TRC_DEBUG("Patching the AoR for %s", _uri.c_str());
+
+  for (BindingPair b : po.get_update_bindings())
+  {
+    TRC_DEBUG("Updating the binding %s", b.first.c_str());
+
+    for (BindingPair b2 : _bindings)
+    {
+      if (b.first == b2.first)
+      {
+        _bindings.erase(b2.first);
+        delete b2.second;
+        break;
+      }
+    }
+
+    Binding* copy_b = new Binding(*(b.second));
+    _bindings.insert(std::make_pair(b.first, copy_b));
+  }
+
+  for (std::string b_id : po.get_remove_bindings())
+  {
+    TRC_DEBUG("Removing the binding %s", b_id.c_str());
+
+    for (BindingPair b : _bindings)
+    {
+      if (b_id == b.first)
+      {
+        _bindings.erase(b.first);
+        delete b.second;
+        break;
+      }
+    }
+  }
+
+  for (SubscriptionPair s : po.get_update_subscriptions())
+  {
+    TRC_DEBUG("Updating the subscription %s", s.first.c_str());
+
+    for (SubscriptionPair s2 : _subscriptions)
+    {
+      if (s.first == s2.first)
+      {
+        _subscriptions.erase(s2.first);
+        delete s2.second;
+        break;
+      }
+    }
+
+    Subscription* copy_s = new Subscription(*(s.second));
+    _subscriptions.insert(std::make_pair(s.first, copy_s));
+  }
+
+  for (std::string s_id : po.get_remove_subscriptions())
+  {
+    TRC_DEBUG("Removing the subscription %s", s_id.c_str());
+
+    for (SubscriptionPair s : _subscriptions)
+    {
+      if (s_id == s.first)
+      {
+        _subscriptions.erase(s.first);
+        delete s.second;
+        break;
+      }
+    }
+  }
+
+  if (po.get_associated_uris())
+  {
+    TRC_DEBUG("Updating the Associated URIs");
+    _associated_uris = po.get_associated_uris().get();
+  }
+
+  if (po.get_increment_cseq())
```

Comment:
I'm struggling to work out what this is doing, and as far as I can tell there are no comments on it anywhere